### ISSUE TITLE
Crystal String behaves inconsistent (and can cause compile time errors)

### ISSUE THREAD

**Issue by wolfgang371** [crystal-lang/crystal#10132]
I stumbled across a bug when reading a .png via `read_file` macro. I cut it down to the following:

```crystal
p "\x0F" # -> "\u000F" (ok)
p "\x83\x0F" # -> "\x83\u000F" (ok)
p "\xB0\x83\x0F" # -> "\xB0\x83\u000F" (ok)
p "\xFE\xB0\x83\x0F" # -> "\xFE\xB0\x83\xF" (inconsistent), triggers "Error: invalid hex escape" if fed as input (e.g. via read_file)
```

I use Crystal 0.35.1 on Ubuntu.
If I try the same thing with Ruby (2.5.1p57) it behaves consistent.

I think this is not related to #2886, but I'm not sure.

**Comment by straight-shoota** [comment]
I can't reproduce this, neither local nor https://carc.in/#/r/a5sk

**Comment by straight-shoota** [comment]
Oh, sry I just noticed the condition about reading from file.

**Comment by straight-shoota** [comment]
So this is a reproducing example:
```cr
{% `echo "\xFE\xB0\x83\x0F" > file.txt` %}
{{ read_file "file.txt" }}
```

```
Called macro defined in eval:3:4

Which expanded to:

 > 1 | "\xFE\xB0\x83\xF\n"
                       ^
Error: invalid hex escape
```

**Comment by wolfgang371** [comment]
yes, that's the compile time error; but the root cause seems to be in the String handling (and I needed the four bytes for it to show up)

**Comment by straight-shoota** [comment]
This seems to be fixed on master. But no idea what did it.
Can you confirm that your use case works with a nightly build?
Or provide a reproducible sample outside the compiler?

**Comment by wolfgang371** [comment]
yes, it does :)
The code now also has the same output as in Ruby.
Thanks!

**Comment by straight-shoota** [comment]
FTR: The fix happened in #9713, thanks @asterite =)

**Comment by wolfgang371** [comment]
maybe one last minor thing to this: `"\xF"` (still) is considered an "invalid hex escape" in Crystal; but it's fine in Ruby
